wom/src/main/scala/wom/util/YamlUtils.scala

```diff
@@ -0,0 +1,102 @@
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.circe.{Json, ParsingFailure}
+import net.ceedubs.ficus.Ficus._
+import net.ceedubs.ficus.readers.ValueReader
+
+import scala.collection.JavaConverters._
+
+object YamlUtils {
+
+  private[util] implicit val refinedNonNegativeReader: ValueReader[Int Refined NonNegative] = {
+    (config: Config, path: String) => {
+      val int = config.getInt(path)
+      refineV[NonNegative](int) match {
+        case Left(error) => throw new BadValue(path, error)
+        case Right(refinedInt) => refinedInt
+      }
+    }
+  }
+
+  private val defaultMaxNodes = ConfigFactory.load().as[Int Refined NonNegative]("yaml.max-nodes")
+
+  /**
+    * Parses the yaml, detecting loops, and a maximum number of nodes.
+    *
+    * See: https://en.wikipedia.org/wiki/Billion_laughs_attack
+    *
+    * @param yaml     The yaml string.
+    * @param maxNodes Maximum number of yaml nodes to parse.
+    * @return The parsed yaml.
+    */
+  def parse(yaml: String, maxNodes: Int Refined NonNegative = defaultMaxNodes): Either[ParsingFailure, Json] = {
+    try {
+      val snakeYamlParser = new org.yaml.snakeyaml.Yaml()
+      val parsed = snakeYamlParser.load[AnyRef](new StringReader(yaml))
+      if (parsed == null) {
+        // Behave just like circe-yaml 0.9.0
+        // https://github.com/circe/circe-yaml/blob/v0.9.0/src/main/scala/io/circe/yaml/parser/package.scala#L79-L98
+        throw new MatchError(parsed)
+      }
+      searchForLoops(parsed, Nil, maxNodes, Array[Int](0))
+      io.circe.yaml.parser.parse(yaml)
+    } catch {
+      case matchError: MatchError =>
+        throw matchError
+      case exception: Exception =>
+        Left(ParsingFailure(exception.getMessage, exception))
+    }
+  }
+
+  /**
+    * Looks for loops and large documents in yaml parsed by SnakeYaml.
+    *
+    * Possibly can be refactored if/when Circe switches to SnakeYaml-Engine (aka Yaml 1.2), as SY-E disables recursive
+    * key by default:
+    *
+    * - https://github.com/circe/circe-yaml/issues/46
+    * - https://bitbucket.org/asomov/snakeyaml/issues/432/aggressive-yaml-anchors-causing
+    * - https://bitbucket.org/asomov/snakeyaml-engine/commits/7573a8b4d551fc84521f4ac1234a361bfbc96698
+    *
+    * @param node     Current node to be evaluated.
+    * @param seen     Previously seen nodes during this branch of cycle checking.
+    * @param maxNodes The maximum number of nodes allowed to be traversed.
+    * @param counter  A "pointer" to the total number of nodes traversed during the entire traversal.
+    */
+  private def searchForLoops(node: AnyRef,
+                             seen: List[AnyRef],
+                             maxNodes: Int Refined NonNegative,
+                             counter: Array[Int]): Unit = {
+    if (seen.exists(_ eq node)) {
+      throw new IllegalArgumentException("Loop detected")
+    }
+
+    counter(0) += 1
```

Comment:
Does this work? Seems at the very least kind of confusing to me.... 

Copying from https://stackoverflow.com/questions/6896118/basic-scala-oop-question-pass-by-reference, could we define a simple `Field` class for this:
```
class Field[T]( var value: T )
```

Reply:
Alternatively, if all you're doing is `get` and `increment`, I've (mis?)used `AtomicInteger` as a similar counter-by-reference value elsewhere